presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalysis.java

```diff
@@ -19,96 +19,102 @@
+    private final Map<NodeRef<Expression>, FieldId> columnReferences;
+    private final Set<NodeRef<InPredicate>> subqueryInPredicates;
+    private final Set<NodeRef<SubqueryExpression>> scalarSubqueries;
+    private final Set<NodeRef<ExistsPredicate>> existsSubqueries;
+    private final Set<NodeRef<QuantifiedComparisonExpression>> quantifiedComparisons;
     // For lambda argument references, maps each QualifiedNameReference to the referenced LambdaArgumentDeclaration
-    private final IdentityLinkedHashMap<Identifier, LambdaArgumentDeclaration> lambdaArgumentReferences;
+    private final Map<NodeRef<Identifier>, LambdaArgumentDeclaration> lambdaArgumentReferences;
 
     public ExpressionAnalysis(
-            IdentityLinkedHashMap<Expression, Type> expressionTypes,
-            IdentityLinkedHashMap<Expression, Type> expressionCoercions,
-            Set<InPredicate> subqueryInPredicates,
-            Set<SubqueryExpression> scalarSubqueries,
-            Set<ExistsPredicate> existsSubqueries,
-            IdentityLinkedHashMap<Expression, FieldId> columnReferences,
-            Set<Expression> typeOnlyCoercions,
-            Set<QuantifiedComparisonExpression> quantifiedComparisons,
-            IdentityLinkedHashMap<Identifier, LambdaArgumentDeclaration> lambdaArgumentReferences)
+            Map<NodeRef<Expression>, Type> expressionTypes,
+            Map<NodeRef<Expression>, Type> expressionCoercions,
+            Set<NodeRef<InPredicate>> subqueryInPredicates,
+            Set<NodeRef<SubqueryExpression>> scalarSubqueries,
+            Set<NodeRef<ExistsPredicate>> existsSubqueries,
+            Map<NodeRef<Expression>, FieldId> columnReferences,
+            Set<NodeRef<Expression>> typeOnlyCoercions,
+            Set<NodeRef<QuantifiedComparisonExpression>> quantifiedComparisons,
+            Map<NodeRef<Identifier>, LambdaArgumentDeclaration> lambdaArgumentReferences)
     {
-        this.expressionTypes = requireNonNull(expressionTypes, "expressionTypes is null");
-        this.expressionCoercions = requireNonNull(expressionCoercions, "expressionCoercions is null");
-        this.typeOnlyCoercions = requireNonNull(typeOnlyCoercions, "typeOnlyCoercions is null");
-        this.columnReferences = new IdentityLinkedHashMap<>(requireNonNull(columnReferences, "columnReferences is null"));
-        this.subqueryInPredicates = requireNonNull(subqueryInPredicates, "subqueryInPredicates is null");
-        this.scalarSubqueries = requireNonNull(scalarSubqueries, "subqueryInPredicates is null");
-        this.existsSubqueries = requireNonNull(existsSubqueries, "existsSubqueries is null");
-        this.quantifiedComparisons = requireNonNull(quantifiedComparisons, "quantifiedComparisons is null");
-        this.lambdaArgumentReferences = requireNonNull(lambdaArgumentReferences, "lambdaArgumentReferences is null");
+        this.expressionTypes = new LinkedHashMap<>(requireNonNull(expressionTypes, "expressionTypes is null"));
```

Comment:
wrap with `unmodifiableSet` in constructor and then just return field reference from `get*` methods.